### PR TITLE
[Pruning, sparsity 2 by 4 documentation]

### PR DESCRIPTION
**Pull request by wwwind**
This PR adds 

- description of the structural pruning (M by N) to the pruning overview document
- a small tutorial that demonstrates how this structural pruning should be used

The tutorial uses a simple cnn model on MNIST dataset, where `Cond2D` layer and `Dense` layers are pruned with structural pruning. The model is converted to TFLite file and weights of these layers are visualize to demonstrate the pattern of structural pruning 2 by 4.

### CODE REVIEW

**Comment by wwwind** [comment]
Hi @daverim and @rino20 As I mentioned before, this PR provides documentation for structural pruning M/N. Please take a look. Thanks!

**Comment by rino20** [comment]
Hi @wwwind, Thanks for working on this. 

I think the new tutorial's content is pretty much overlapped with the existing guide - pruning_comprehensive_guide. So what about adding a subsection in the existing guide instead of adding a new page? It could better elaborate the difference between random vs structured vs block sparsity. 

Also, it seems better to merge the experiment results to the existing table in the index page. 

Please let me know what you think about this.

**Comment by wwwind** [comment]
Hi @rino20 
The model is defined differently in the comprehesnsive guide + I need to do conversion to tflite to demonstrate 2/4 sparsity, because only tflite file has this pattern. I can move this guide to the comprehensive_guide, but in my opinion the comprehensive guide will be overloaded. I would prefer personally to have a separate tutorial similar to the tutorial pruning_for_on_device_inference. But if you think that we should incorporate this into comprehensive guide than we can do this.

Regarding the table with results:
We have slightly different numbers (baseline, 50%) for InceptionV3 and MV2. Should we correct the original numbers ?

**Comment by rino20** [comment]
Although it is not desirable to add dedicated page for every techniques we add, I think it can be justified due to the tutorial's complexity and we have no good documentation structure to incorporate it nicely. Let's have separate page for 2x4 tutorial. 

On the other hand, the metric table needs to be merged, it would be super weird to have two different table with different numbers in same setting. If the baseline is different, please try to replicate it.

**Comment by wwwind** [comment]
Hi @rino20 PR has been updated per your suggestion. Please take a look. Thanks

File: tensorflow_model_optimization/g3doc/guide/pruning/index.md [hunk 1]
```diff
@@ -41,6 +43,18 @@ It is on our roadmap to add support in the following areas:
 *   [Minimal Subclassed model support](https://github.com/tensorflow/model-optimization/issues/155)
 *   [Framework support for latency improvements](https://github.com/tensorflow/model-optimization/issues/173)
 
+## Structural pruning M by N
```
Comment (rino20) [inline tensorflow_model_optimization/g3doc/guide/pruning/index.md:46]: The overview page is not for a specific pruning technique. This section doesn't look very appropriate to me being here, so what about moving this to the subpage (tutorial) if you want to keep it

Rather than this section, I think the experiment results / and a brief comment below the result about what 2x4 means / would be enough.

File: tensorflow_model_optimization/g3doc/guide/pruning/index.md [hunk 2]
```diff
@@ -51,12 +65,14 @@ It is on our roadmap to add support in the following areas:
...
       <th>Non-sparse Top-1 Accuracy </th>
       <th>Sparse Accuracy </th>
+      <th>Sparsity 2 by 4</th>
       <th>Sparsity </th>
     </tr>
     <tr>
       <td rowspan=3>InceptionV3</td>
       <td rowspan=3>78.1%</td>
       <td>78.0%</td>
+      <td>75.8%</td>
```
Comment (rino20) [inline tensorflow_model_optimization/g3doc/guide/pruning/index.md:75]: The current table is hard to understand. I think we can do either

Sparse accuracy | Sparsity
77%.                    | 50%
69%                    | 2 by 4

or 
Ramdom sparse accuracy | Random sparsity | Structured sparse accuracy | Structured sparsity
77%.                    | 50%.      |   69%                    | 2 by 4


And just FYI, the original numbers for inceptionv3 in this table is from the paper linked below - https://arxiv.org/pdf/1710.01878.pdf

File: tensorflow_model_optimization/g3doc/guide/pruning/pruning_with_sparsity_2_by_4.ipynb [hunk 3]
```diff
@@ -0,0 +1,571 @@
...
+      "source": [
+        "print(f\"Shape of Dense layer is {tensor_data.shape}\")"
+      ],
+      "outputs": [],
+      "metadata": {}
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "Now we visualize the structure for a small subset of the weight tensor. The structure of the weight tensor is sparse in the last dimension and has a pattern (2,4): two elements out of four are zeros. To make visualization more clear, we replace all non-zero values with ones."
```
[the remark this answers is not in the document]
Comment (wwwind) [reply tensorflow_model_optimization/g3doc/guide/pruning/pruning_with_sparsity_2_by_4.ipynb:366]: Personally, I would keep visualization of Dense layer. As this is important to demonstrate that this pruning is actual for these two types of layers and in which dimension Dense layer is pruned.

**Comment by wwwind** [comment]
Hi @rino20 Thanks for the review. I addressed your comments. Please take a look

File: tensorflow_model_optimization/g3doc/guide/pruning/index.md [hunk 4]
```diff
@@ -113,6 +122,28 @@ The models were tested on Imagenet.
...
+
+DS-CNN-L is a keyword spotting model created for edge devices. It can be found
+in [Arm’s ML Examples repository](https://github.com/ARM-software/ML-examples/tree/master/tflu-kws-cortex-m).
+
+<figure>
+  <table>
+    <tr>
+      <th>Model</th>
+      <th>Unpruned</th>
+      <th>Sparsity 2 by 4 </th>
```
Comment (rino20) [inline tensorflow_model_optimization/g3doc/guide/pruning/index.md:135]: Could you update these naming to be aligned with above? Thanks.

File: tensorflow_model_optimization/g3doc/guide/pruning/pruning_with_sparsity_2_by_4.ipynb [hunk 5]
```diff
@@ -0,0 +1,597 @@
...
+      ],
+      "metadata": {}
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "Structural pruning zeroes out model weights at the beginning of the training\n",
+        "process according to the following pattern: M weights are set to zero in the\n",
+        "block of N weights. It is important to notice that this pattern affects only the last dimension of the weight tensor for the model that is converted by TensorFlow Lite. For example, `Conv2D` layer weights in TensorFlow Lite have the structure [channel_out, height, width, channel_in] and `Dense` layer weights have the structure [channel_out, channel_in]. The sparsity pattern is applied to the weights in the last dimension: channel_in.\n",
+        "Special hardware can benefit from this type of sparsity in the model and inference time can have a speedup up to 2x. Because this pattern lock in sparsity is more restrictive, the accuracy achieved after fine-tuning is worse than with the magnitude-based pruning.\n",
```
Comment (rino20) [inline tensorflow_model_optimization/g3doc/guide/pruning/pruning_with_sparsity_2_by_4.ipynb:96]: 2x speedup is for 1x2 or 2x4 sparsity, so I prefer omitting a specific number for the speedup.

File: tensorflow_model_optimization/g3doc/guide/pruning/pruning_with_sparsity_2_by_4.ipynb [hunk 6]
```diff
@@ -0,0 +1,597 @@
...
+      "cell_type": "markdown",
+      "source": [
+        "There is a python script included in the TensorFlow Model Optimization Toolkit that could be used to check whether which layers in the model from the given flite file have the structurally pruned weights: [`check_sparsity_m_by_n.py`](https://github.com/tensorflow/model-optimization/blob/master/tensorflow_model_optimization/python/core/sparsity/keras/tools/check_sparsity_m_by_n.py). The usage of this tool for the case of 2 by 4 is shown below:"
+      ],
+      "metadata": {}
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "`python ./tensorflow_model_optimization/python/core/sparsity/keras/tools/check_sparsity_m_by_n.py --model_tflite=pruned_model.tflite --m_by_n=2,4`"
```
Comment (rino20) [inline tensorflow_model_optimization/g3doc/guide/pruning/pruning_with_sparsity_2_by_4.ipynb:562]: Could you format it to a shell script as you did above in Setup section? e.g, "! pip install ... "

**Comment by wwwind** [comment]
Hi @rino20 Thanks for the review. I have pushed corrections. Please take a look.

**Review by rino20 — APPROVED** [review]

**Comment by wwwind** [comment]
Hi @rino20 I see some internal checks are failing for this PR. Could you please take a look ? What should be corrected ? Thanks

**Comment by rino20** [comment]
> Hi @rino20 I see some internal checks are failing for this PR. Could you please take a look ? What should be corrected ? Thanks

No worries, I fixed the errors internally. After some internal reviews the change will be published.

File: tensorflow_model_optimization/g3doc/guide/pruning/pruning_with_sparsity_2_by_4.ipynb [hunk 7]
```diff
@@ -0,0 +1,599 @@
...
+      "metadata": {}
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "Structural pruning zeroes out model weights at the beginning of the training\n",
+        "process according to the following pattern: M weights are set to zero in the\n",
+        "block of N weights. It is important to notice that this pattern affects only the last dimension of the weight tensor for the model that is converted by TensorFlow Lite. For example, `Conv2D` layer weights in TensorFlow Lite have the structure [channel_out, height, width, channel_in] and `Dense` layer weights have the structure [channel_out, channel_in]. The sparsity pattern is applied to the weights in the last dimension: channel_in.\n",
+        "Special hardware can benefit from this type of sparsity in the model and inference time can have a significant speedup. Because this pattern lock in sparsity is more restrictive, the accuracy achieved after fine-tuning is worse than with the magnitude-based pruning.\n",
+        "It is important to indicate that the pattern is valid only for the model that is converted to tflite.\n",
```
Comment (rino20) [inline tensorflow_model_optimization/g3doc/guide/pruning/pruning_with_sparsity_2_by_4.ipynb:97]: HI @wwwind, 

While internal review, I found I don't understand this sentence. Could you explain it with reply?

Comment (wwwind) [reply tensorflow_model_optimization/g3doc/guide/pruning/pruning_with_sparsity_2_by_4.ipynb:97]: Hi @rino20 

If we check the model that is not saved to tflite format, then we don't see this pattern in the last dimension. 
Because, for Conv2D, for example: internally in TF data format we have (h, w, channel_in(kernel_size), channel_out(filters)). when we save into the tflite data format, we have: (channel_out, h, w, channel_in). The sparsity pattern is in the last dimension.
